### PR TITLE
Gor

### PR DESCRIPTION
**Pull request by spongezhang**
Make several changes to your code. 
1. Removing input_norm(), it makes no sense in test phase
2. Adding support for gor from "Learning Spread-out Local Feature Descriptors. X Zhang, FX Yu, S Kumar, SF Chang, ICCV 2017".  It further reduces FPR95 by 10% relatively on UBC dataset. 
3. Moving all the code file to ./code/ subdirectory 
4. Supporting tensorboard log with different hyperparameters
5. Separating log directory and model directory

### CODE REVIEW

**Comment by ducha-aiki** [comment]
Hi @spongezhang , thanks for contributing a lot!

@DagnyT will check your PR, I only need to point out, that  per-patch contrast normalization  is indeed very important part during test time. 
E.g. it is shown in https://arxiv.org/abs/1504.06603, Figure 3 (1st row vs 2nd) it work also used in L2Net , and out own experience confirms its importance, especially if one change BatchNorm + ReLU to the ELU for the speed-up.

**Comment by spongezhang** [comment]
Interesting! I would check the paper and details. Based on my experience on image retrieval, it will greatly downgrades the performance. For example, assuming we have two exactly the same patches in two different batches, if we do input norm, the two same patches will have different descriptors. It's a little bit against the nature of the descriptor extraction.

**Comment by ducha-aiki** [comment]
@spongezhang you are right in that example :)
That is why we extract mean of each patch from itself.  input_norm do following:
batch of patches 128 x 1 x 32 x 32  -> 128 x 1 x 1 x 1 (128 means for each patch) -> expand to original size -> subtract.

**Comment by spongezhang** [comment]
Thank you for the explanation. It makes sense when images have lightening conditions. I'll double check and send you a new pull request.
